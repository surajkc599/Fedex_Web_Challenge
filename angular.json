{
  "version": 1,
  "projects": {
    "giffy-image-search": {
      "projectType": "application",
      "schematics": {
        "@nrwl/angular:component": {
          "style": "scss"
        }
      },
      "root": "apps/giffy-image-search",
      "sourceRoot": "apps/giffy-image-search/src",
      "prefix": "fedex",
      "architect": {
        "build": {
          "builder": "@angular-devkit/build-angular:browser",
          "options": {
            "outputPath": "dist/apps/giffy-image-search",
            "index": "apps/giffy-image-search/src/index.html",
            "main": "apps/giffy-image-search/src/main.ts",
            "polyfills": "apps/giffy-image-search/src/polyfills.ts",
            "tsConfig": "apps/giffy-image-search/tsconfig.app.json",
            "aot": true,
            "assets": [
              "apps/giffy-image-search/src/favicon.ico",
              "apps/giffy-image-search/src/assets"
            ],
            "styles": ["apps/giffy-image-search/src/styles.scss"],
            "stylePreprocessorOptions": {
              "includePaths": ["libs/foundation/src/lib"]
            },
            "extractCss": true,
            "scripts": []
          },
          "configurations": {
            "production": {
              "fileReplacements": [
                {
                  "replace": "apps/giffy-image-search/src/environments/environment.ts",
                  "with": "apps/giffy-image-search/src/environments/environment.prod.ts"
                }
              ],
              "optimization": true,
              "outputHashing": "all",
              "sourceMap": false,
              "extractCss": true,
              "namedChunks": false,
              "extractLicenses": true,
              "vendorChunk": false,
              "buildOptimizer": true,
              "budgets": [
                {
                  "type": "initial",
                  "maximumWarning": "2mb",
                  "maximumError": "5mb"
                },
                {
                  "type": "anyComponentStyle",
                  "maximumWarning": "6kb",
                  "maximumError": "10kb"
                }
              ]
            }
          }
        },
        "serve": {
          "builder": "@angular-devkit/build-angular:dev-server",
          "options": {
            "browserTarget": "giffy-image-search:build"
          },
          "configurations": {
            "production": {
              "browserTarget": "giffy-image-search:build:production"
            }
          }
        },
        "extract-i18n": {
          "builder": "@angular-devkit/build-angular:extract-i18n",
          "options": {
            "browserTarget": "giffy-image-search:build"
          }
        },
        "lint": {
          "builder": "@angular-devkit/build-angular:tslint",
          "options": {
            "tsConfig": [
              "apps/giffy-image-search/tsconfig.app.json",
              "apps/giffy-image-search/tsconfig.spec.json"
            ],
            "exclude": ["**/node_modules/**", "!apps/giffy-image-search/**/*"]
          }
        },
        "test": {
          "builder": "@nrwl/jest:jest",
          "options": {
            "jestConfig": "apps/giffy-image-search/jest.config.js",
            "tsConfig": "apps/giffy-image-search/tsconfig.spec.json",
            "passWithNoTests": true,
            "setupFile": "apps/giffy-image-search/src/test-setup.ts"
          }
        }
      }
    },
    "giffy-image-search-e2e": {
      "root": "apps/giffy-image-search-e2e",
      "sourceRoot": "apps/giffy-image-search-e2e/src",
      "projectType": "application",
      "architect": {
        "e2e": {
          "builder": "@nrwl/cypress:cypress",
          "options": {
            "cypressConfig": "apps/giffy-image-search-e2e/cypress.json",
            "tsConfig": "apps/giffy-image-search-e2e/tsconfig.e2e.json",
            "devServerTarget": "giffy-image-search:serve"
          },
          "configurations": {
            "production": {
              "devServerTarget": "giffy-image-search:serve:production"
            }
          }
        },
        "lint": {
          "builder": "@angular-devkit/build-angular:tslint",
          "options": {
            "tsConfig": ["apps/giffy-image-search-e2e/tsconfig.e2e.json"],
            "exclude": [
              "**/node_modules/**",
              "!apps/giffy-image-search-e2e/**/*"
            ]
          }
        }
      }
    },
    "foundation": {
      "projectType": "library",
      "root": "libs/foundation",
      "sourceRoot": "libs/foundation/src",
      "prefix": "fedex",
      "architect": {
        "lint": {
          "builder": "@angular-devkit/build-angular:tslint",
          "options": {
            "tsConfig": [
              "libs/foundation/tsconfig.lib.json",
              "libs/foundation/tsconfig.spec.json"
            ],
            "exclude": ["**/node_modules/**", "!libs/foundation/**/*"]
          }
        },
        "test": {
          "builder": "@nrwl/jest:jest",
          "options": {
            "jestConfig": "libs/foundation/jest.config.js",
            "tsConfig": "libs/foundation/tsconfig.spec.json",
            "passWithNoTests": true,
            "setupFile": "libs/foundation/src/test-setup.ts"
          }
        }
      },
      "schematics": {
        "@nrwl/angular:component": {
          "style": "scss"
        }
      }
    },
    "table": {
      "projectType": "library",
      "root": "libs/table",
      "sourceRoot": "libs/table/src",
      "prefix": "fedex",
      "architect": {
        "lint": {
          "builder": "@angular-devkit/build-angular:tslint",
          "options": {
            "tsConfig": [
              "libs/table/tsconfig.lib.json",
              "libs/table/tsconfig.spec.json"
            ],
            "exclude": ["**/node_modules/**", "!libs/table/**/*"]
          }
        },
        "test": {
          "builder": "@nrwl/jest:jest",
          "options": {
            "jestConfig": "libs/table/jest.config.js",
            "tsConfig": "libs/table/tsconfig.spec.json",
            "passWithNoTests": true,
            "setupFile": "libs/table/src/test-setup.ts"
          }
        }
      },
      "schematics": {
        "@nrwl/angular:component": {
          "style": "scss"
        }
      }
    },
    "paginator": {
      "projectType": "library",
      "root": "libs/paginator",
      "sourceRoot": "libs/paginator/src",
      "prefix": "fedex",
      "architect": {
        "lint": {
          "builder": "@angular-devkit/build-angular:tslint",
          "options": {
            "tsConfig": [
              "libs/paginator/tsconfig.lib.json",
              "libs/paginator/tsconfig.spec.json"
            ],
            "exclude": ["**/node_modules/**", "!libs/paginator/**/*"]
          }
        },
        "test": {
          "builder": "@nrwl/jest:jest",
          "options": {
            "jestConfig": "libs/paginator/jest.config.js",
            "tsConfig": "libs/paginator/tsconfig.spec.json",
            "passWithNoTests": true,
            "setupFile": "libs/paginator/src/test-setup.ts"
          }
        }
      },
      "schematics": {
        "@nrwl/angular:component": {
          "style": "scss"
        }
      }
    },
    "data": {
      "root": "libs/data",
      "sourceRoot": "libs/data/src",
      "projectType": "library",
      "schematics": {},
      "architect": {
        "lint": {
          "builder": "@angular-devkit/build-angular:tslint",
          "options": {
            "tsConfig": [
              "libs/data/tsconfig.lib.json",
              "libs/data/tsconfig.spec.json"
            ],
            "exclude": ["**/node_modules/**", "!libs/data/**/*"]
          }
        },
        "test": {
          "builder": "@nrwl/jest:jest",
          "options": {
            "jestConfig": "libs/data/jest.config.js",
            "tsConfig": "libs/data/tsconfig.spec.json",
            "passWithNoTests": true
          }
        }
      }
    },
    "textbox": {
      "projectType": "library",
      "root": "libs/textbox",
      "sourceRoot": "libs/textbox/src",
      "prefix": "fedex",
      "architect": {
        "lint": {
          "builder": "@angular-devkit/build-angular:tslint",
          "options": {
            "tsConfig": [
              "libs/textbox/tsconfig.lib.json",
              "libs/textbox/tsconfig.spec.json"
            ],
            "exclude": ["**/node_modules/**", "!libs/textbox/**/*"]
          }
        },
        "test": {
          "builder": "@nrwl/jest:jest",
          "options": {
            "jestConfig": "libs/textbox/jest.config.js",
            "tsConfig": "libs/textbox/tsconfig.spec.json",
            "passWithNoTests": true,
            "setupFile": "libs/textbox/src/test-setup.ts"
          }
        }
      },
      "schematics": {
        "@nrwl/angular:component": {
          "style": "scss"
        }
      }
    },
    "loader": {
      "projectType": "library",
      "root": "libs/loader",
      "sourceRoot": "libs/loader/src",
      "prefix": "fedex",
      "architect": {
        "lint": {
          "builder": "@angular-devkit/build-angular:tslint",
          "options": {
            "tsConfig": [
              "libs/loader/tsconfig.lib.json",
              "libs/loader/tsconfig.spec.json"
            ],
            "exclude": ["**/node_modules/**", "!libs/loader/**/*"]
          }
        },
        "test": {
          "builder": "@nrwl/jest:jest",
          "options": {
            "jestConfig": "libs/loader/jest.config.js",
            "tsConfig": "libs/loader/tsconfig.spec.json",
            "passWithNoTests": true,
            "setupFile": "libs/loader/src/test-setup.ts"
          }
        }
      },
      "schematics": {
        "@nrwl/angular:component": {
          "style": "scss"
        }
      }
    }
  },
  "cli": {
    "defaultCollection": "@nrwl/angular"
  },
  "schematics": {
    "@nrwl/workspace": {
      "library": {
        "linter": "tslint"
      }
    },
    "@nrwl/cypress": {
      "cypress-project": {
        "linter": "tslint"
      }
    },
    "@nrwl/node": {
      "application": {
        "linter": "tslint"
      },
      "library": {
        "linter": "tslint"
      }
    },
    "@nrwl/nest": {
      "application": {
        "linter": "tslint"
      },
      "library": {
        "linter": "tslint"
      }
    },
    "@nrwl/express": {
      "application": {
        "linter": "tslint"
      },
      "library": {
        "linter": "tslint"
      }
    },
    "@nrwl/angular:application": {
      "unitTestRunner": "jest",
      "e2eTestRunner": "cypress"
    },
    "@nrwl/angular:library": {
      "unitTestRunner": "jest"
    }
  },
  "defaultProject": "giffy-image-search"
}
