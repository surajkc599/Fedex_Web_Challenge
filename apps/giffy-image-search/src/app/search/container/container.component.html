<!--Container that wraps Header, searchbox, table and paginator-->
<div class="container" fxLayout="column" fxLayoutGap="20px">
  <!--Textbox where user can enter text-->
  <fedex-textbox fxFlex="5%">
    <input
      #input
      role="textbox"
      type="text"
      placeholder="Search..."
      [ngModel]="searchText$ | async"
      (ngModelChange)="searchText$.next(input.value)"
    />
  </fedex-textbox>
  <!--Show Table, Pagination and Validation of messages-->
  <ng-container *ngIf="loadingState$ | async as loading">
    <!--Show validation of messages-->
    <fedex-message [message]="loading"></fedex-message>
    <!--Show table if results are available-->
    <ng-container *ngIf="datasource$ | async as results">
      <ng-container
        *ngIf="results.data.length && loading.state !== 'profanity'"
      >
        <!--Table that displays the results-->
        <fedex-table [datasource]="results.data" [columns]="columns" fxFlex>
          <ng-template fedexTemplate="header">
            <tr>
              <th *ngFor="let column of columns" class="fh-primary">
                {{ column }}
              </th>
            </tr>
          </ng-template>
          <ng-template fedexTemplate="body" let-gif>
            <tr>
              <td *ngFor="let column of columns" class="fb-primary">
                <ng-template [ngIf]="column === 'image'" [ngIfElse]="default">
                  <img
                    [src]="gif?.images?.original?.url"
                    alt="image-loaded"
                    [style.width.px]="200"
                    [style.height.px]="200"
                  />
                </ng-template>
                <ng-template #default>
                  {{ gif[column] }}
                </ng-template>
              </td>
            </tr>
          </ng-template>
        </fedex-table>
        <!--Pagination-->
        <fedex-paginator
          [recordsPerPage]="25"
          [totalRecords]="results?.pagination?.total_count"
          (pageChange)="onPageChange($event)"
          fxFlex="10%"
        ></fedex-paginator>
      </ng-container>
    </ng-container>
  </ng-container>
</div>
