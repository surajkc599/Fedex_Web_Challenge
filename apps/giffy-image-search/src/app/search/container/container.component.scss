@import 'spacing/variables';

// Way of over-riding base variable
$spacing-05: 40px;

:host{
  width: 100%;
  height: 100%;

  > .container {
    width: inherit;
    height: inherit;
    padding: 0 $spacing-05;
  }
}
