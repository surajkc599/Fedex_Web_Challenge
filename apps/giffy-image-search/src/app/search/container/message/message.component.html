<ng-container [ngSwitch]="message?.state">
  <!-- SHOW WHEN THE TEXT ENTERED IS PROFANITY-->
  <fedex-loader class="loader" [loading]="true" *ngSwitchCase="'loading'"></fedex-loader>
  <!-- SHOW WHEN THE TEXT ENTERED IS PROFANITY-->
  <p class="fb-primary" *ngSwitchCase="'profanity'">
    Please enter valid text. We are not obliged to diplay results for profanity
    word...
  </p>
  <!--SHOW WHEN EMPTY RESULTS LOADED-->
  <ng-container class="fb-primary" *ngSwitchCase="'empty'">
    <p class="fb-primary">
      Please enter valid text. Search text did not find any results.
    </p>
  </ng-container>
  <!--SHOW WHEN ERROR-->
  <ng-container class="fb-primary" *ngSwitchCase="'error'">
    <p class="fb-primary">
      Something went wrong. Please try again...
    </p>
  </ng-container>
</ng-container>
