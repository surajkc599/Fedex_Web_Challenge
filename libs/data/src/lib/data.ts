// This could be also part of application itself.
// Keeping this under libs, would help to share same interface between application and libs and may be also with Backend like NodeJS/NESTJS
// Because of Monorepo we can have all apps in one space.

export enum ContentState {
  LOADING = 'loading',
  LOADED = 'loaded',
  ERROR = 'error',
  EMPTY = 'empty',
  PROFANITY = 'profanity'
}

export interface IGiffySearch {
  data: IGiffyData[];
  pagination: IPagination;
}

export interface IGiffyData {
  images: IOriginalImage;
  type: string;
  title: string;
  id: string;
  rating: string;
}

export interface IOriginalImage {
  original: {
    url: string;
    width: string;
    height: string;
  };
}

export interface IPagination {
  total_count: number;
  offset: number;
  count: number;
}

export interface IPaginator {
  offset: number;
  endIndex: number;
}

export interface IContentState {
  state: ContentState;
  error?: boolean;
}
