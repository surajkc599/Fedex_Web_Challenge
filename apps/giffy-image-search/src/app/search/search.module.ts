import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';

import { AppFlexLayoutModule } from '../flex-module/flex-layout.module';

// Internal table module exposed from libs/table
import { TableModule } from '@fedex/table';
import { TextboxModule } from '@fedex/textbox';
import { PaginatorModule } from '@fedex/paginator';
import { LoaderModule } from '@fedex/loader';

import { SearchRoutingModule } from './search.routing.module';
import { ContainerComponent } from './container/container.component';
import { MessageComponent } from './container/message/message.component';

@NgModule({
  declarations: [ContainerComponent, MessageComponent],
  imports: [
    CommonModule,
    FormsModule,
    SearchRoutingModule,
    AppFlexLayoutModule,
    TableModule,
    TextboxModule,
    PaginatorModule,
    LoaderModule
  ],
})
export class SearchModule {}
