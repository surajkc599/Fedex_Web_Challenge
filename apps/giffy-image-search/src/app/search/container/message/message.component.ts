import { Component, OnInit, Input } from '@angular/core';
import { IContentState } from '@fedex/data';

/**
 * Component that is responsible for showing messages.
 * 1. If empty response
 * 2. If the text is profanity
 * 3. If error occurs
 * 4. show loader
 */
@Component({
  selector: 'fedex-message',
  templateUrl: './message.component.html',
  styleUrls: ['./message.component.scss'],
})
export class MessageComponent {
  @Input() message: IContentState;
}
