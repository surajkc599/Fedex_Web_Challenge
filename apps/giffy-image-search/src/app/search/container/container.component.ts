import { Component, OnInit, OnDestroy, AfterViewInit } from '@angular/core';
import {
  Observable,
  combineLatest,
  BehaviorSubject,
  of,
} from 'rxjs';
import {
  debounceTime,
  distinctUntilChanged,
  switchMap,
  map,
  catchError,
  startWith,
  tap,
} from 'rxjs/operators';

import {
  IPaginator,
  IContentState,
  ContentState,
  IGiffySearch,
} from '@fedex/data';

import { SearchService } from '../services/search.service';
import { ProfanityService } from '../services/profanity.service';

const DEBOUNCE_TIME = 500;

/**
 * Container that wraps the search feature.
 *
 * This combines searchbox and its corresponding results.
 * The results are represented in table format along with pagination.
 *
 * Based on Comments -> Compodoc can generate readable static site.
 */
@Component({
  selector: 'fedex-search-container',
  templateUrl: './container.component.html',
  styleUrls: ['./container.component.scss'],
})
export class ContainerComponent implements OnInit, AfterViewInit, OnDestroy {
  // Options for Table component
  columns = ['image', 'type', 'title', 'id', 'rating'];
  datasource$: Observable<IGiffySearch>;

  // Observable for search text
  searchText$ = new BehaviorSubject<string>('');
  // Observable for paginator change
  paginatorChange$ = new BehaviorSubject<IPaginator>({
    endIndex: 0,
    offset: 0,
  });
  // Internal observable for view to represent loading state
  loadingState$: Observable<IContentState>;

  /**
   * @ignore
   */
  constructor(
    private searchService: SearchService,
    private profanityService: ProfanityService
  ) {}

  /**
   * Initialization
   * Datasource gets binded async from search service -> search store
   *
   * Emit default puppies text so user can some results to see on page load.
   */
  ngOnInit() {
    this.datasource$ = this.searchService.getGifsData();
    this.searchText$.next('puppies');
  }

  /**
   * Responsible for couple of activities after view is loaded
   *
   * 1. Listen for changes in textbox
   * 2. Listen for changes in paginator
   * 3. combines both observables and reacts when either of gets triggered
   * 4. Check for profanity for the entered text
   * 5. If not, then load the gifss data
   * 6. switchMap cancels the previous subscription when new value passes through after debounce time.
   *
   *  No need of unsubscription coz of binding in template directly
   *
   *  Returns loading state based on the activity happened
   */
  ngAfterViewInit() {
    this.loadingState$ = combineLatest([
      this.searchText$,
      this.paginatorChange$,
    ]).pipe(
      debounceTime(DEBOUNCE_TIME),
      distinctUntilChanged(),
      switchMap(([text, paginator]) =>
        this.containsProfanity(text).pipe(
          switchMap((isProfanity) =>
            isProfanity
              ? of({ state: ContentState.PROFANITY })
              : this.loadGifsData(text, paginator.offset).pipe(
                  map(({ data }) => {
                    if (data.length) {
                      return { state: ContentState.LOADED };
                    } else {
                      return { state: ContentState.EMPTY };
                    }
                  })
                )
          ),
          startWith({
            state: ContentState.LOADING,
          })
        )
      ),
      catchError(() =>
        of({
          state: ContentState.ERROR,
          error: true,
        })
      )
    );
  }

  /**
   *
   * @param event Emitted from Paginator component
   */
  onPageChange(event: IPaginator) {
    this.paginatorChange$.next(event);
  }

  /**
   *
   * @param text to process for profanity check
   *
   * @returns observable of boolean value
   */
  containsProfanity(text) {
    return this.profanityService.containsProfanity(text);
  }

  /**
   *
   * @param text Entered text by ser
   * @param offset offset to load gifs data for the entered text
   *
   * @returns observable of IGiffySearch
   */
  loadGifsData(text: string, offset: number) {
    return this.searchService.loadAndGetObservableGifsData(text, offset);
  }

  ngOnDestroy() {
    this.searchText$.unsubscribe();
    this.paginatorChange$.unsubscribe();
  }
}
