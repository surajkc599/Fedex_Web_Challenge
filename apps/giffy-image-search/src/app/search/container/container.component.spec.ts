import { render, waitFor } from '@testing-library/angular';
import { TestBed } from '@angular/core/testing';
import { Injectable } from '@angular/core';
import { of } from 'rxjs';

import { FormsModule } from '@angular/forms';
import { TableModule } from '@fedex/table';
import { TextboxModule } from '@fedex/textbox';
import { PaginatorModule } from '@fedex/paginator';
import { LoaderModule } from '@fedex/loader';
import { ScrollingModule } from '@angular/cdk/scrolling';

import { ContainerComponent } from './container.component';
import { MessageComponent } from './message/message.component';

import { SearchService } from '../services/search.service';
import { ProfanityService } from '../services/profanity.service';

import { IGiffySearch } from '@fedex/data';

let mockProfanityService: ProfanityService;
let mockSearchService: SearchService;

const mockGifsData: IGiffySearch = {
  pagination: {
    total_count: 2,
    count: 2,
    offset: 0,
  },
  data: [
    {
      images: {
        original: {
          url: 'http://somedomain/test/gif/car1.gif',
          width: '200',
          height: '200',
        },
      },
      type: 'gif',
      title: 'Car-1',
      id: '1',
      rating: 'A',
    },
    {
      images: {
        original: {
          url: 'http://somedomain/test/gif/car2.gif',
          width: '200',
          height: '200',
        },
      },
      type: 'gif',
      title: 'Car-2',
      id: '2',
      rating: 'B',
    },
  ],
};

@Injectable()
class MockSearchService {
  getGifsData() {
    return of(mockGifsData);
  }
  loadAndGetObservableGifsData() {
    return of(mockGifsData);
  }
}

@Injectable()
class MockProfanityService {
  containsProfanity() {
    return of(false);
  }
}

jest.useFakeTimers();

describe('Search feature Container - Valid Text', () => {
  function setup() {
    return render(ContainerComponent, {
      componentProperties: {},
      declarations: [ContainerComponent, MessageComponent],
      providers: [
        ContainerComponent,
        { provide: SearchService, useClass: MockSearchService },
        { provide: ProfanityService, useClass: MockProfanityService },
      ],
      imports: [
        FormsModule,
        TableModule,
        TextboxModule,
        PaginatorModule,
        LoaderModule,
        ScrollingModule,
      ],
    });
  }

  it('should capture the initial state of search feature', async () => {
    const { fixture } = await setup();
    expect(fixture).toMatchSnapshot();
  });

  it('should display records for valid search text', async (done) => {
    const { fixture, getByText } = await setup();

    // Arrange
    const text = 'cars';
    mockProfanityService = TestBed.inject(ProfanityService);
    mockSearchService = TestBed.inject(SearchService);

    // Enter cars as text
    fixture.componentInstance.searchText$.next('');
    fixture.componentInstance.paginatorChange$.next({ offset: 0, endIndex: 0 });
    fixture.detectChanges();
    fixture.autoDetectChanges();

    // Advance becasue of DEBOUNCE TIME
    jest.advanceTimersByTime(600);

    fixture.componentInstance.searchText$.next(text);
    fixture.detectChanges();
    fixture.autoDetectChanges();

    waitFor(() => {
      expect(getByText('Car-1')).toBeDefined();
      done();
    });
  });
});

@Injectable()
class StubProfanityService {
  containsProfanity() {
    return of(true);
  }
}

@Injectable()
class StubSearchService {
  getGifsData() {
    return of(mockGifsData);
  }
  loadAndGetObservableGifsData() {
    return of(mockGifsData);
  }
}
describe('Search feature Container - Profanity Text', () => {
  function setup() {
    return render(ContainerComponent, {
      componentProperties: {},
      declarations: [ContainerComponent, MessageComponent],
      providers: [
        ContainerComponent,
        { provide: SearchService, useClass: StubSearchService },
        { provide: ProfanityService, useClass: StubProfanityService },
      ],
      imports: [
        FormsModule,
        TableModule,
        TextboxModule,
        PaginatorModule,
        LoaderModule,
        ScrollingModule,
      ],
    });
  }

  it('should display validation message for profanity word', async () => {
    const { fixture, getByText } = await setup();

    // Arrange
    const text = 'fuck';
    mockProfanityService = TestBed.inject(ProfanityService);
    mockSearchService = TestBed.inject(SearchService);

    // Enter cars as text
    fixture.componentInstance.searchText$.next('');
    fixture.componentInstance.paginatorChange$.next({ offset: 0, endIndex: 0 });
    fixture.detectChanges();

    // Advance becasue of DEBOUNCE TIME
    jest.advanceTimersByTime(600);

    fixture.componentInstance.searchText$.next(text);
    fixture.detectChanges();

    const expectation = `Please enter valid text. We are not obliged to diplay results for profanity word...`;

    // Assert
    expect(getByText(expectation)).toBeDefined();
  });
});
