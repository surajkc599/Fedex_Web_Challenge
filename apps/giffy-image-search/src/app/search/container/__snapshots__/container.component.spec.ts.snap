// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search feature Container - Valid Text should capture the initial state of search feature 1`] = `
<fedex-search-container
  columns={[Function Array]}
  datasource$={[Function Observable]}
  loadingState$={[Function Observable]}
  paginatorChange$={[Function BehaviorSubject]}
  profanityService={[Function MockProfanityService]}
  searchService={[Function MockSearchService]}
  searchText$={[Function BehaviorSubject]}
>
  <div
    class="container"
    fxlayout="column"
    fxlayoutgap="20px"
  >
    <fedex-textbox
      fxflex="5%"
    >
      <section
        class="layout-container"
      >
        <input
          class="ng-untouched ng-pristine ng-valid"
          placeholder="Search..."
          role="textbox"
          type="text"
        />
      </section>
    </fedex-textbox>
    
  </div>
</fedex-search-container>
`;
